solutions/usaco-694.mdx

```diff
@@ -104,5 +104,90 @@ int main() {
 ```
 
 </CPPSection>
+	
+<JavaSection>
+	
+```java
+import java.io.*;
+import java.util.*;
+
+public class HoofPaperScissors {
+	static int[] moves;
+
+	public static void main (String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new FileReader("hps.in"));
+		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("hps.out")));
+
+		StringTokenizer inputLine = new StringTokenizer(br.readLine());
+		int numGames = Integer.parseInt(inputLine.nextToken());
+		int maximumSwitches = Integer.parseInt(inputLine.nextToken());
+		// dp[game][switches][type] = maximum number of wins
+		// at that game with at most that amount
+		// of switches and that with that type active
+		moves = new int[numGames + 1];
+		for (int x = 1; x <= numGames; x++) {
+			char c = br.readLine().charAt(0);
+			// set H as 0, P as 1, S as 2
+			if (c == 'H') {
+				moves[x] = 0;
+			}
+			else if (c == 'P') {
+				moves[x] = 1;
+			}
+			else {
+				moves[x] = 2;
+			}
+		}
+
+		int[][][] dp = new int[numGames + 1][maximumSwitches + 2][3];
+		int maximum = 0;
+
+		/*
+		c = current game number
+		n = number of switches
+		 */
+
+		for (int c = 1; c <= numGames; c++) {  // for each game
+			// for each switch amount
+			for (int n = 1; n < maximumSwitches + 2; n++) {
+				// try switching to the three possibilities
+				int add1 = 0;
+				if (moves[n] == 0) {
+					add1 = 1;
+				}
+				// take the maximum of the game before + add (win or lose)
+				// and switching from the other two options
+				dp[c][n][0] = Math.max(dp[c - 1][n][0]
+				+ add1, Math.max(dp[c - 1][n - 1][1] + add1,
+				dp[c - 1][n - 1][2] + add1));
+
+				int add2 = 0;
+				if (moves[c] == 1) {
+					add2 = 1;
+				}
+				dp[c][n][1] = Math.max(dp[c - 1][n][1]
+				+ add2, Math.max(dp[c - 1][n - 1][0] + add2,
+				dp[c - 1][n - 1][2] + add2));
```

Comment:
```suggestion
				dp[c][n][1] = Math.max(
								dp[c - 1][n][1] + add2, 
								Math.max(
									dp[c - 1][n - 1][0] + add2,
									dp[c - 1][n - 1][2] + add2
								)
							);
```
i'm not sure; does this look better?

Reply:
i don't think so, that looks kind of weird